### PR TITLE
Integration test harness: shared scenario catalog, live repo testing, interactive shell

### PR DESCRIPTION
**Pull request by nodeselector**
The integration tests were a loose collection of Ruby scripts with scenarios defined inline.
This PR replaces that with a structured, cross-language scenario catalog and an interactive
shell for rapid edit-build-test cycles.

## What changed

**Shared scenario catalog** (`test/scenarios/catalog.yml` + `catalog.go`)

A single YAML file defines all 36 scenarios across 9 categories (happy path, edge cases,
lockfile, SSO, monorepo, etc.). Both the Go test suite and the Ruby integration harness
consume the same catalog, so scenario definitions stay in sync. The Go side gets a typed
`Scenario` struct with `LoadCatalog()` and tests to validate the YAML parses correctly.

**Live repo testing**

Five live scenarios clone real repos (github/github, vercel/next.js, github/internal-actions,
github/mcv3-boot, github/launch) and run the binary against their actual workflows. Auth
falls back to `gh auth token` when no env var is set. The shell also supports ad-hoc repo
targeting -- just type `run owner/repo` or bare `owner/repo` to test any GitHub repo on
the fly without predefining it in the catalog.

**Interactive shell (REPL)**

`make test-shell` drops into a comfortable development shell with:

- Tab completion and history via Reline (proper macOS support, no EditLine quirks)
- `run <name>`, `run <owner/repo>`, `run all`, bare name shortcuts
- `build` rebuilds the binary without leaving the shell
- `inspect`, `cd`, `diff`, `rerun` for exploring scenario fixtures
- `profile [dir|off]` for CPU/trace profiling
- `pause` toggle for stepping through run-all
- `auth`, `status`, `clear`, `help`
- Ctrl+C handling at every level (prompt, mid-run, run-all loop)
- PTY output with chunked reads to avoid flicker

## Testing

- `go test ./...` passes (including new catalog_test.go)
- All 15 stub scenarios pass in batch mode
- Live scenarios tested against all 5 target repos
- Shell manually tested for completion, Ctrl+C/D, all commands

## Related branches (stacked)

These are separate PRs stacked on this branch:

- `nodeselector/cli-ui-fixes` -- 5 CLI terminal UX fixes
- `nodeselector/fix-spinner-flicker` -- buffered frame writes + DEC mode 2026
- `nodeselector/lab-workflows-dir-override` -- `GH_ACTIONS_PIN_WORKFLOWS_DIR` env var

### CODE REVIEW

File: test/integration/harness.rb [hunk 1]
```diff
@@ -124,9 +129,19 @@ def stop
       end
     end
 
+    # ── Catalog ─────────────────────────────────────────────────────────
+    # Loads the shared scenario catalog from test/scenarios/catalog.yml.
+    class Catalog
+      def self.load
+        catalog_path = File.expand_path("../../scenarios/catalog.yml", __FILE__)
+        YAML.load_file(catalog_path)
```
[the remark this answers is not in the document]
Comment (nodeselector) [reply test/integration/harness.rb]: Switched to `YAML.safe_load_file` with `permitted_classes: [Symbol]`.

File: test/integration/harness.rb [hunk 2]
```diff
@@ -270,13 +301,57 @@ def prepare(binary)
...
-        ctx = prepare(binary)
+      def run(binary, profile_dir: nil)
+        # Live repo scenarios need auth — explicit token or gh CLI
+        if @live_repo
+          token = ENV["GH_TOKEN"] || ENV["GITHUB_TOKEN"]
+          if token.nil? || token.empty?
+            gh_token = `gh auth token 2>/dev/null`.strip
+            raise SkipScenario, "no GH_TOKEN or gh auth" if gh_token.empty?
+          end
+        end
```
[the remark this answers is not in the document]
Comment (nodeselector) [reply test/integration/harness.rb:358]: Added `@needs_token` flag to `Scenario` and check it alongside `@live_repo` in `Scenario#run` -- auth is enforced before execution for any scenario that declares it needs a token.

File: test/integration/run.rb [hunk 3]
```diff
@@ -90,334 +85,336 @@ def sso_403_all(srv)
...
-        { "Content-Type" => "application/json" },
-        JSON.generate({ message: "Not Found" })
-      ]
+  # Token-required scenarios skip gracefully without a token
+  if needs_token
+    s.assert_custom do |r|
+      token = ENV["GH_TOKEN"] || ENV["GITHUB_TOKEN"] || ""
+      nil if token.empty? # skip
     end
   end
```
[the remark this answers is not in the document]
Comment (nodeselector) [reply test/integration/run.rb:165]: The assertion-based skip is replaced with `s.needs_token(true)` which sets a flag on the Scenario. Auth is now enforced in `Scenario#run` before execution via `raise SkipScenario`.

File: test/integration/harness.rb [hunk 4]
```diff
@@ -357,12 +433,19 @@ def run_pty
         begin
           PTY.spawn("/bin/bash", "-c", shell_cmd) do |reader, _writer, pid|
             begin
-              reader.each_char do |ch|
-                $stdout.print ch       # live to terminal
-                combined << ch
+              # Read in chunks so we flush once per available burst instead
+              # of once per byte. Eliminates flicker while keeping spinners
+              # responsive (readpartial returns as soon as data is available).
+              $stdout.sync = true
```
[the remark this answers is not in the document]
Comment (nodeselector) [reply test/integration/harness.rb:446]: Now saves `prev_sync = $stdout.sync` before setting true and restores it in ensure.

File: test/integration/harness.rb [hunk 5]
```diff
@@ -357,12 +433,19 @@ def run_pty
...
+              loop do
+                chunk = reader.readpartial(4096)
+                $stdout.write chunk
+                combined << chunk
               end
-            rescue Errno::EIO
+            rescue Errno::EIO, EOFError
               # PTY closed — normal on macOS when process exits
+            ensure
+              $stdout.sync = false
```
[the remark this answers is not in the document]
Comment (nodeselector) [reply test/integration/harness.rb]: Same fix -- the ensure block now restores the saved `prev_sync` value.

File: test/integration/harness.rb [hunk 6]
```diff
@@ -547,17 +706,151 @@ def shell
...
+
+          when "clear"
+            print "\033[2J\033[H"
+
+          when "build"
+            repo_root = File.expand_path("../..", __dir__)
+            print "  \e[2mbuilding…\e[0m "
+            $stdout.flush
+            t0 = Process.clock_gettime(Process::CLOCK_MONOTONIC)
+            ok = system("cd #{Shellwords.shellescape(repo_root)} && go build -o gh-actions-pin ./cmd/gh-actions-pin 2>&1")
```
[the remark this answers is not in the document]
Comment (nodeselector) [reply test/integration/harness.rb]: Switched to `system("go", "build", "-o", "gh-actions-pin", "./cmd/gh-actions-pin", chdir: repo_root)` -- argv array, no shell.

File: Makefile [hunk 7]
```diff
@@ -7,7 +7,7 @@ EXT_DIR := $(XDG_DATA_HOME)/gh/extensions/$(EXT_NAME)
 
 RUBY := $(shell command -v /opt/homebrew/opt/ruby/bin/ruby 2>/dev/null || echo ruby)
 
-.PHONY: build test test-integration test-shell install reinstall uninstall
+.PHONY: build test test-integration test-shell test-live test-matrix install reinstall uninstall
```
[the remark this answers is not in the document]
Comment (nodeselector) [reply Makefile]: Added `test-smoke`, `test-stub`, and `test-real` to the `.PHONY` line.

File: .github/workflows/test.yml [hunk 8]
```diff
@@ -19,7 +19,7 @@ jobs:
       GONOSUMDB: github.com/github/*
     steps:
       - name: OIDC Setup for goproxy
-        uses: github/setup-goproxy@5e60e1074d42316dfe2949ebf9a92bf77b24645b # v1.1.0
+        uses: github/setup-goproxy@v1.1.0
```
[the remark this answers is not in the document]
Comment (nodeselector) [reply .github/workflows/test.yml]: Pinned to `5e60e1074d42316dfe2949ebf9a92bf77b24645b # v1.1.0`.

File: .github/workflows/release.yml [hunk 9]
```diff
@@ -13,7 +13,7 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - name: OIDC Setup for goproxy
-        uses: github/setup-goproxy@5e60e1074d42316dfe2949ebf9a92bf77b24645b # v1.1.0
+        uses: github/setup-goproxy@v1.1.0
```
[the remark this answers is not in the document]
Comment (nodeselector) [reply .github/workflows/release.yml]: Pinned to `5e60e1074d42316dfe2949ebf9a92bf77b24645b # v1.1.0`.

File: test/integration/run.rb [hunk 10]
```diff
@@ -90,334 +85,336 @@ def sso_403_all(srv)
...
-  s.assert_failure
-  # The "resolution failed:" wrapper should NOT appear in output
-  s.assert_output_excludes("resolution failed:")
-  # The clean error should appear
-  s.assert_output_contains("SAML enforcement")
+tag_filter = nil
+tag_filter = "live" if ARGV.delete("--live")
+tag_filter = "stub" if ARGV.delete("--stub")
+tag_filter = "smoke" if ARGV.delete("--smoke")
+tag_filter = "real_repo" if ARGV.delete("--real")
```
[the remark this answers is not in the document]
Comment (nodeselector) [reply test/integration/run.rb]: Now collects all tag flags and rejects with a clear error if more than one is provided.
